src/graph/dg.rs

```diff
@@ -374,71 +374,191 @@ pub fn resolve_all_paths_only_of_references(
+//             path.extend(stack_graph, &mut paths, &mut queue);
+//         }
+//     }
+//     progress_bar.finish();
+//     log::info!("found {} bindings", bindings);
+// }
+
+pub fn resolve_all_paths_reference_by_reference(
     explorer: &mut SynchroExplorer,
     stack_graph: &StackGraph
 ) {
     let mut references = Vec::<Handle<Node>>::new();
     log::info!("finding references");
     for node_handle in stack_graph.iter_nodes() {
-        match stack_graph.index(node_handle) {
-            PushSymbol(_) => {
-                if stack_graph.index(node_handle).is_reference() {
-                    let refkind = Refkind::from(find_debug_info(
-                        stack_graph,
-                        node_handle,
-                        "refkind".to_string()
-                    ).unwrap_or_default());
-                    match refkind.is_nothing() {
+        if stack_graph.index(node_handle).is_reference() {
+            let refkind = Refkind::from(find_debug_info(
+                stack_graph,
+                node_handle,
+                "refkind".to_string()
+            ).unwrap_or_default());
+            match refkind.is_nothing() {
+                false => {
+                    references.push(node_handle);
+                },
+                true => (),
+            }
+        }
+    }
+    log::info!("found {} references", references.len());
+    let mut bindings = 0;
+    let progress_bar = indicatif::ProgressBar::new(references.len().try_into().unwrap());
+    for reference in references {
+        let mut partials = stack_graphs::partial::PartialPaths::new();
+        ForwardPartialPathStitcher::find_all_complete_partial_paths(
+            &mut GraphEdgeCandidates::new(stack_graph, &mut partials, None),
+            std::iter::once(reference),
+            &NoCancellation,
+            |sg,_ps,p| {
+                if p.is_complete(sg) {
+                    match Defkind::from(
+                        find_debug_info(
+                            stack_graph,
+                            p.end_node,
+                            "defkind".to_string()
+                        ).unwrap_or_default()
+                    ).is_nothing() {
+                        true => {},
                         false => {
-                            references.push(node_handle);
-                        },
-                        true => (),
+                            if explorer.name_bindings.get(&p.start_node).is_none() {
+                                bindings += 1;
+                                progress_bar.inc(1);
+                                explorer.set_name_binding(p.start_node, p.end_node);
+                            }
+                        }
                     }
                 }
-            },
-            _ => {}
+            }
+        ).unwrap();
+    }
+    progress_bar.finish();
+    log::info!("found {} bindings", bindings);
+}
+
+pub fn resolve_all_paths_with_partial_paths(
```

Comment:
This uses an intermediate steps where partial paths are precomputed. I hoped that would help with the performance for the large JUnit case, but I didn't see much effect yet.